sound/soc/sof/intel/hda-pcm.c

```diff
@@ -95,6 +95,83 @@ static inline u32 get_bits(struct snd_sof_dev *sdev, int sample_bits)
 	}
 };
 
+/* There are two dma modes in dsp:
+ * (1) host dma and link dma work in decouple mode for SOF + HDA
+ * (2) host dma and GP dma work in couple mode for SOF + I2S
+ * The difference is that in decopule mode, the host dma is set by dai
+ * frontend and link dma is set by dai backend and the setting maybe
+ * different while in couple mode, host dma and GP dma share the same
+ * dma setting.
+ */
+static inline int get_be_dma_channel(struct snd_sof_dev *sdev,
+				     struct snd_pcm_substream *substream,
+				     struct sof_ipc_stream_params *ipc_params)
+{
+	struct snd_soc_pcm_runtime *fe = substream->private_data;
+	struct hdac_ext_stream *be_stream = NULL;
+	struct snd_soc_dpcm *dpcm;
+	int direction = substream->stream;
+	struct sof_intel_be_params *be_params;
+
+	be_params = (struct sof_intel_be_params *)ipc_params->private_data;
```

Comment:
why do you need private data here?
I am confused on what you are trying to do.

Reply:
Liam suggested that: hda is special to intel and not generic thing. If sony or other vendors want to use SOF, they don't like link_dma_channel, and they may also have some special setting to send to FW, so I discussed with team members about this, and decide to reserved a private_data for this condition.

Reply:
use a union instead if you are concerned about code size or non-common parameters.